src/components/storage/index.ts

```diff
@@ -334,3 +352,128 @@ export class IpfsStorage extends Storage {
     return await encryptData(response.data, encryptionType)
   }
 }
+
+export class S3Storage extends Storage {
+  public constructor(file: S3FileObject) {
+    super(file)
+    const [isValid, message] = this.validate()
+    if (isValid === false) {
+      throw new Error(`Error validationg the S3 file: ${message}`)
+    }
+  }
+
+  validate(): [boolean, string] {
+    const file: S3FileObject = this.getFile() as S3FileObject
+    if (!file.hash) {
+      return [false, 'Missing Hash']
+    }
+    return [true, '']
+  }
+
+  parseDecryptedStream(decryptedStream: Readable): Promise<S3Object> {
+    return new Promise((resolve, reject) => {
+      let data = ''
+      decryptedStream.on('data', (chunk) => {
+        data += chunk
+      })
+      decryptedStream.on('end', () => {
+        try {
+          const parsedData = JSON.parse(data)
+          resolve(parsedData)
+        } catch (error) {
+          reject(error)
+        }
+      })
+      decryptedStream.on('error', (error) => {
+        reject(error)
+      })
+    })
+  }
+
+  getDownloadUrl(): string {
+    const fileHash = this.getFile().hash
+    return fileHash
+  }
+
+  async getS3Object(): Promise<S3Object> {
+    const file = this.getFile()
+    const fileStream = Readable.from(file.hash)
+    const streamString = await streamToString(fileStream)
+    const encryptedData = ethers.getBytes(streamString)
+    const decryptedData = await decryptData(encryptedData, file.encryptMethod)
+    return JSON.parse(decryptedData.toString()) as S3Object
+  }
+
+  async fetchData(): Promise<any> {
+    const s3Obj = await this.getS3Object()
+    const spacesEndpoint = new AWS.Endpoint(s3Obj.endpoint)
+    const s3 = new AWS.S3({
+      endpoint: spacesEndpoint,
+      accessKeyId: s3Obj.accessKeyId,
+      secretAccessKey: s3Obj.secretAccessKey,
+      region: s3Obj.region
+    })
+
+    const params = {
+      Bucket: s3Obj.bucket,
+      Key: s3Obj.objectKey
+    }
+    try {
+      const data = await s3.getObject(params).promise()
```

Comment:
is there any way of making this to work with streams?
Downloading/encrypting/decrypting 1TB file will not work with promises & streamToString

Reply:
we will investigate

Reply:
https://medium.com/@v1shva/uploading-downloading-files-using-aws-s3-sdk-v3-javascript-typescript-as-a-stream-and-processing-42ba07bb892c

Reply:
i updated the code. I used the link provided as a reference